native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeObjCRefinementChecker.kt

```diff
@@ -0,0 +1,185 @@
+    }
+
+    private fun DeclarationDescriptor.findRefinedAnnotations(): Pair<List<AnnotationDescriptor>, List<AnnotationDescriptor>> {
+        val objCAnnotations = mutableListOf<AnnotationDescriptor>()
+        val swiftAnnotations = mutableListOf<AnnotationDescriptor>()
+        for (annotation in annotations) {
+            val annotations = annotation.annotationClass?.annotations ?: continue
+            for (metaAnnotation in annotations) {
+                when (metaAnnotation.fqName) {
+                    refinesForObjCFqName -> {
+                        objCAnnotations.add(annotation)
+                        break
+                    }
+                    refinesInSwiftFqName -> {
+                        swiftAnnotations.add(annotation)
+                        break
+                    }
+                }
+            }
+        }
+        return objCAnnotations to swiftAnnotations
+    }
+
+    private fun checkOverrides(
+        declaration: KtDeclaration,
+        descriptor: CallableMemberDescriptor,
+        context: DeclarationCheckerContext,
+        objCAnnotations: List<AnnotationDescriptor>,
+        swiftAnnotations: List<AnnotationDescriptor>
+    ) {
+        if (descriptor.overriddenDescriptors.isEmpty()) return
+        var isRefinedForObjC = objCAnnotations.isNotEmpty()
+        var isRefinedInSwift = swiftAnnotations.isNotEmpty()
+        val supersNotRefinedForObjC = mutableListOf<CallableMemberDescriptor>()
+        val supersNotRefinedInSwift = mutableListOf<CallableMemberDescriptor>()
+        for (overriddenDescriptor in descriptor.overriddenDescriptors) {
+            val (superIsRefinedForObjC, superIsRefinedInSwift) = overriddenDescriptor.inheritsRefinedAnnotations()
+            if (superIsRefinedForObjC) isRefinedForObjC = true else supersNotRefinedForObjC.add(overriddenDescriptor)
+            if (superIsRefinedInSwift) isRefinedInSwift = true else supersNotRefinedInSwift.add(overriddenDescriptor)
+        }
+        if (isRefinedForObjC && supersNotRefinedForObjC.isNotEmpty()) {
+            context.trace.reportIncompatibleOverride(declaration, objCAnnotations, supersNotRefinedForObjC)
+        }
+        if (isRefinedInSwift && supersNotRefinedInSwift.isNotEmpty()) {
+            context.trace.reportIncompatibleOverride(declaration, swiftAnnotations, supersNotRefinedInSwift)
+        }
+    }
+
+    private fun checkFakeOverrides(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        if (descriptor !is ClassDescriptor) return
+        descriptor.defaultType.memberScope
+            .getContributedDescriptors(DescriptorKindFilter.ALL, MemberScope.Companion.ALL_NAME_FILTER)
+            .forEach {
+                if (it !is CallableMemberDescriptor || it.kind.isReal) return@forEach
+                checkOverrides(declaration, it, context, emptyList(), emptyList())
```

Comment:
Shouldn't we also update the `INCOMPATIBLE_OBJC_REFINEMENT_OVERRIDE` message, just like in the other PR?